File: Java/Civitas/src/civitas/Tablero.java
package civitas;
import java.util.ArrayList;

public class Tablero {
    private int numCasillaCarcel;
    private ArrayList<Casilla_Descanso> casillas;
    private int porSalida;
    private boolean tieneJuez;
    
    public Tablero(int tNumCasillaCarcel){
        numCasillaCarcel = tNumCasillaCarcel;
        porSalida = 0;
        tieneJuez = false;
        casillas = new ArrayList<>();
    }
    
    private boolean correcto(){
        return casillas.size() > numCasillaCarcel;
    }
    
    private boolean correcto(int numCasilla){
        return correcto() && (casillas.size() > numCasilla);
    }
    
    public int getCarcel(){
        return numCasillaCarcel;
    }
    
    public int getPorSalida(){
        int t = porSalida;
        if (t != 0){
            porSalida--;
        }
        return t;
    }
    
    public void añadeCasilla(Casilla_Descanso casilla){
        casillas.add(casilla);
        if (casillas.size() == numCasillaCarcel){
            casillas.add(new Casilla_Descanso("Carcel"));
        }
    }
    
    public void añadeJuez(){
        if (!tieneJuez){
            añadeCasilla(new Casilla_Juez(numCasillaCarcel,"Juez"));
            tieneJuez = true;
        }
    }
    
    public Casilla_Descanso getCasilla(int numCasilla){
        return (numCasilla >= 0) && (numCasilla < casillas.size()) ? casillas.get(numCasilla) : null;
    }
    
    public int nuevaPosicion(int actual, int tirada){
        if(!correcto()){
            return -1;
        }
        int ncasilla = (actual + tirada) % casillas.size();
        if(ncasilla != actual + tirada){
            porSalida++;
        }
        return ncasilla;
    }
    
    public int calcularTirada(int origen, int destino){
        return (destino-origen+casillas.size())%casillas.size();
    }
    
}

File: Ruby/CivitasR/lib/Tablero.rb
#encoding:utf-8
require_relative 'Casilla'
require_relative 'Enum'
require_relative "MazoSorpresas"

module Civitas
  class Tablero
    
    
    def initialize(carcel)
      @numCasillaCarcel= carcel >= 1 ? carcel : 1
      @casillas=[Casilla_Descanso.new("Salida")]
      @porSalida=0
      @tieneJuez=false
    end
    
    def correcto(numCasilla = -1)
       return @tieneJuez && @casillas.length > @numCasillaCarcel && numCasilla < @casillas.length
    end
    
    private :correcto

    def getPorSalida
      xp=@porSalida
      if(@porSalida>0)
        @porSalida=@porSalida-1
      end
      return xp
    end

    def añadeCasilla(casilla)
      if(@casillas.length==@numCasillaCarcel)
        @casillas.push(Casilla_Descanso.new("Carcel"))
      end
      @casillas.push(casilla)
      if(@casillas.length==@numCasillaCarcel)
        @casillas.push(Casilla_Descanso.new("Carcel"))
      end
    end

    def añadeJuez
      if(!@tieneJuez)  
        añadeCasilla(Casilla_Juez.new(@numCasillaCarcel,"Juez"))
        @tieneJuez=true
      end
    end

    def getCasilla(numCasilla)
      return correcto(numCasilla) ? @casillas[numCasilla] : nil
    end

    def nuevaPosicion(actual,tirada)
      resultado=actual+tirada
      if(correcto(actual))
        if((actual+tirada)>=@casillas.length)
          resultado%=@casillas.length
          @porSalida+=1
        end
      else
        resultado=-1
      end
      return resultado
    end

    def calcularTirada(origen,destino)
      (destino-origen)>0 ? resul=destino-origen : resul=@casillas.length+(destino-origen)
      return resul
    end
    
    attr_reader :casillas,:numCasillaCarcel
    
  end
end

